kernel/sched.c

```diff
@@ -1347,3 +1378,78 @@ int k_thread_cpu_mask_disable(k_tid_t thread, int cpu)
 }
 
 #endif /* CONFIG_SCHED_CPU_MASK */
+
+int z_impl_k_thread_join(struct k_thread *thread, s32_t timeout)
+{
+	k_spinlock_key_t key;
+	int ret;
+
+	__ASSERT(((arch_is_in_isr() == false) || (timeout == K_NO_WAIT)), "");
+
+	key = k_spin_lock(&sched_spinlock);
+
+	if ((thread->base.thread_state & _THREAD_DEAD) != 0) {
+		ret = 0;
+		goto out;
+	}
+
+	if (timeout == K_NO_WAIT) {
+		ret = -EBUSY;
+		goto out;
+	}
+
+	if (thread->base.pended_on == &_current->base.join_waiters) {
+		ret = -EDEADLK;
+		goto out;
+	}
```

Comment:
Wouldn't better make this check early ? Thread A may call it with K_NO_WAIT in a loop trying to join thread B (that has already join thread A). In this situation thread A will always receive EBUSY and will keep trying to join B and never finish, causing a deadlock since thread B is waiting A to finish.
 I mean, EDADLK is a problem that the caller must to handle while EBUSY is something that says to the caller to try later, so the first should have high priority.

Reply:
Yeah I can move this up.

Reply:
moved